### PR TITLE
NIAD-1976 testing fixes for medication mapper

### PR DESCRIPTION
**Pull request by ElliotB-96**
Updates to medication request mapper based on testing feedback

### CODE REVIEW

File: schema/src/main/java/org/hl7/v3/RCMRMT030101UK04Authorise.java [hunk 1]
```diff
@@ -327,6 +327,17 @@ public boolean hasQuantity() {
         return this.predecessor;
     }
 
+    public RCMRMT030101UK04Predecessor getPredecessorFirstRep() {
```
Comment (rorysproule2) [inline schema/src/main/java/org/hl7/v3/RCMRMT030101UK04Authorise.java:330]: Does this Authorise class and the Predecessor class need added to the manually edited classes in the schema README?

**Review by dawidskainos — CHANGES_REQUESTED** [review]
I'm afraid we can't use class static fields for any state keeping (MEDICATION_IDS). This is not thread safe. This Map needs to exist only in the context of a single mapping call. So it has to be either a variable created in scope of a method or we need to use something like ThreadLocal.

File: schema/src/main/java/org/hl7/v3/RCMRMT030101UK04Authorise.java (hunk 1 above)
Comment (ElliotB-96) [reply schema/src/main/java/org/hl7/v3/RCMRMT030101UK04Authorise.java:330]: Authorise is already in the readme, added Predecessor.

**Review by rorysproule2 — APPROVED** [review]

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java [hunk 2]
```diff
@@ -90,4 +95,8 @@ private static boolean hasManufacturedMaterial(RCMRMT030101UK04Consumable consum
         return consumable.hasManufacturedProduct() && consumable.getManufacturedProduct().hasManufacturedMaterial()
             && consumable.getManufacturedProduct().getManufacturedMaterial().hasCode();
     }
+
+    protected static void resetMedicationMaps() {
```
Comment (dawidskainos) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java:99]: why protected? there's no inheritance here

**Review by rorysproule2 — DISMISSED** [review]

File: gp2gp-translator/src/integrationTest/resources/application.yml [hunk 3]
```diff
@@ -9,7 +9,7 @@ spring:
   datasource:
     url: ${PS_DB_URL:jdbc:postgresql://localhost:5436}/patient_switching
     username: gp2gp_user
-    password: ${GP2GP_TRANSLATOR_USER_DB_PASSWORD}
+    password: password
```
Comment (rorysproule2) [inline gp2gp-translator/src/integrationTest/resources/application.yml:12]: Revert this pls

**Review by rorysproule2 — APPROVED** [review]

**Review by rorysproule2 — DISMISSED** [review]

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperContext.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        return keyBuilder(CD::hasCode, CD::getCode, code)
+            + keyBuilder(CD::hasOriginalText, CD::getOriginalText, code)
+            + keyBuilder(CD::hasDisplayName, CD::getDisplayName, code);
+    }
+
+    private static String keyBuilder(Function<CD, Boolean> checker, Function<CD, String> getter, CD code) {
+        if (checker.apply(code)) {
+            return getter.apply(code);
+        }
+        return null;
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperContext.java:48]: Maybe just do:
```suggestion
        return checker.apply(code) ? getter.apply(code) : null;
```

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperUtils.java [hunk 5]
```diff
@@ -111,4 +114,31 @@ protected static RCMRMT030101UK04Authorise extractSupplyAuthorise(RCMRMT030101UK
...
+
+    public static Stream<RCMRMT030101UK04MedicationStatement> extractAllMedications(RCMRMT030101UK04Component4 component4) {
+        return Stream.concat(
+            Stream.of(component4.getMedicationStatement()),
+            component4.hasCompoundStatement()
+                ? CompoundStatementUtil.extractResourcesFromCompound(component4.getCompoundStatement(),
+                    RCMRMT030101UK04Component02::hasMedicationStatement, RCMRMT030101UK04Component02::getMedicationStatement)
+                .stream()
+                .map(RCMRMT030101UK04MedicationStatement.class::cast)
+                : Stream.empty()
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperUtils.java:141]: Could you move this fragment to a separate method for readability?

**Review by mrogowska — CHANGES_REQUESTED** [review]

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperUtils.java (hunk 5 above)
Comment (ElliotB-96) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperUtils.java:141]: certainly

**Review by mrogowska — APPROVED** [review]

**Review by rorysproule2 — APPROVED** [review]

**Pull request merged**